### PR TITLE
Add Intel XPU device support

### PR DESCRIPTION
**Pull request by SjeYinTeoIntel**
## Summary

Adds Intel XPU support (Intel GPU via `torch.xpu`) to the device utility module.
Previously, passing `xpu` to any device utility raised `AssertionError`, and
`get_device()` had no XPU branch — Intel GPU hardware silently fell back to CPU.

---

### `modelscope/utils/constant.py`
| Before | After |
|---|---|
| `cpu`, `gpu` only | Added `xpu = 'xpu'` to `Devices` class |

### `modelscope/utils/device.py`

| Function | Change |
|---|---|
| `verify_device` | Added `'xpu'` to accepted allowlist; default `device_id = 0` for bare `'xpu'`; updated docstring |
| `create_device` | Added XPU branch → `torch.device('xpu:N')`; falls back to CPU with log warning |
| `device_placement` | Added XPU branch for PyTorch path → `torch.xpu.set_device()`; falls back silently |
| `get_device` | Extended XPU branch to respect `LOCAL_RANK` for multi-XPU distributed training |

> **Multi-XPU distributed example:** with 4 XPUs and `LOCAL_RANK` set,
> worker 0 → `xpu:0`, worker 1 → `xpu:1`, worker 2 → `xpu:2`, worker 3 → `xpu:3`

All XPU paths guarded with `hasattr(torch, 'xpu') and torch.xpu.is_available()` —
**zero impact on existing CUDA or CPU environments.**

### `tests/utils/test_device.py`

| Test | Behaviour |
|---|---|
| `test_verify` | Added `xpu` and `xpu:1` assertion cases |
| `test_get_device_xpu` | `@skipUnless` XPU available — asserts `device.type == 'xpu'` |
| `test_get_device_cpu` | Skips when accelerator present — asserts CPU fallback |
| `test_create_device_xpu` | `@skipUnless` XPU available — asserts correct `xpu:0` device |
| `test_create_device_xpu_fallback` | Skips when XPU present — asserts CPU fallback |
| `test_device_placement_torch_xpu` | `@skipUnless` XPU available — asserts `torch.xpu.current_device() == 0` |

---

## Test results

> Tested on Intel XPU hardware — `torch==2.12.0+xpu` · `xpu_count=1` · `Python 3.12.3`

<details>
<summary>▶ Click to expand full test output</summary>

platform linux -- Python 3.12.3, pytest-9.1.1, pluggy-1.6.0
rootdir: /home/dut3806/sy/modelscope

tests/utils/test_device.py::DeviceTest::test_create_device_torch          PASSED   [ 10%]
tests/utils/test_device.py::DeviceTest::test_create_device_xpu            PASSED   [ 20%]
tests/utils/test_device.py::DeviceTest::test_create_device_xpu_fallback   SKIPPED  [ 30%]
tests/utils/test_device.py::DeviceTest::test_device_placement_cpu         PASSED   [ 40%]
tests/utils/test_device.py::DeviceTest::test_device_placement_tf_gpu      SKIPPED  [ 50%]
tests/utils/test_device.py::DeviceTest::test_device_placement_torch_gpu   PASSED   [ 60%]
tests/utils/test_device.py::DeviceTest::test_device_placement_torch_xpu   PASSED   [ 70%]
tests/utils/test_device.py::DeviceTest::test_get_device_cpu               SKIPPED  [ 80%]
tests/utils/test_device.py::DeviceTest::test_get_device_xpu               PASSED   [ 90%]
tests/utils/test_device.py::DeviceTest::test_verify                       PASSED   [100%]

========================= 7 passed, 3 skipped =========================

</details>

### Result breakdown

| Status | Count | Tests |
|---|---|---|
| ✅ PASSED | 7 | Includes 3 real XPU hardware assertions |
| ⏭️ SKIPPED | 3 | XPU present → fallback tests skipped; 1 pre-existing TF skip |
| ❌ FAILED | 0 | — |

> **How to read the skips:**
> - `test_create_device_xpu_fallback` and `test_get_device_cpu` are skipped *because* XPU is present — they exist to verify CPU fallback on non-XPU machines
> - On a CPU-only or CUDA machine the pattern inverts: XPU tests skip, fallback tests pass

---

## Environment requirement

> ⚠️ Standard `pip install torch` does **not** include `torch.xpu`.
> Without the XPU wheels, `hasattr(torch, 'xpu')` returns `False` and all XPU
> paths are skipped automatically — no errors, no breakage for existing users.

Install the Intel XPU wheels to enable XPU support:

```bash
pip3 install torch==2.12.0+xpu -index-url https://download.pytorch.org/whl/xpu

### CODE REVIEW

**Comment by SjeYinTeoIntel** [comment]
> This PR is stale because it has been open 45 days with no activity. Remove stale label or comment or this will be closed in 10 days.

May I know how to remove stale? I'm waiting reviewers to response.

**Comment by SjeYinTeoIntel** [comment]
Hi @wangxingjun778 ,

Quick ping on this PR when you have a moment.

It adds Intel XPU device support and utilities to help expand hardware options for the community. I've tested it locally on actual Intel XPU hardware, and all the unit tests in test_device.py are passing successfully.

Let me know if the changes look good to you or if you'd like me to tweak anything. Thanks!

**Comment by SjeYinTeoIntel** [comment]
Hi @wangxingjun778 @yingdachen @tastelikefeet ,

I've opened an issue showing that Intel XPUs aren't fully supported by the framework's device utilities yet: 
https://github.com/modelscope/modelscope/issues/1804

This PR resolves it by adding native XPU backend detection and testing. The implementation is completely isolated from existing CUDA/CPU code paths, so there is zero risk of breaking current device logic. 

This safely opens up ModelScope compatibility for the growing community of Intel XPU users. I'd love it if you could take a look when you have a moment! Thanks! 🙏

**Comment by tastelikefeet** [comment]
> Hi @wangxingjun778 @yingdachen @tastelikefeet ,
> 
> I've opened an issue showing that Intel XPUs aren't fully supported by the framework's device utilities yet: #1804
> 
> This PR resolves it by adding native XPU backend detection and testing. The implementation is completely isolated from existing CUDA/CPU code paths, so there is zero risk of breaking current device logic.
> 
> This safely opens up ModelScope compatibility for the growing community of Intel XPU users. I'd love it if you could take a look when you have a moment! Thanks! 🙏

Sorry for waiting so long, Let me see

**Review by tastelikefeet — APPROVED** [review]

**Comment by SjeYinTeoIntel** [comment]
> > Hi @wangxingjun778 @yingdachen @tastelikefeet ,
> > I've opened an issue showing that Intel XPUs aren't fully supported by the framework's device utilities yet: #1804
> > This PR resolves it by adding native XPU backend detection and testing. The implementation is completely isolated from existing CUDA/CPU code paths, so there is zero risk of breaking current device logic.
> > This safely opens up ModelScope compatibility for the growing community of Intel XPU users. I'd love it if you could take a look when you have a moment! Thanks! 🙏
> 
> Sorry for waiting so long, Let me see

**Comment by SjeYinTeoIntel** [comment]
@tastelikefeet  sorry for that. accidentally press wrong >< reopen this PR.

**Comment by siawchen** [comment]
Hi @tastelikefeet - Thanks for approving the PR. Will you be able to approve the workflows too? Or let us know whom we shall check for the workflows approval? 

Thanks!

**Comment by tastelikefeet** [comment]
Lint failed. Run `pre-commit run --all-files` and commit the file please.

**Comment by SjeYinTeoIntel** [comment]
> pre-commit run --all-files

Checked and resolved. Done updated and commit the file. Please help to trigger the workflow again.

**Comment by SjeYinTeoIntel** [comment]
Thanks @tastelikefeet for the approval 👍
